net/tstun/mtu_test.go

```diff
@@ -4,25 +4,81 @@ package tstun
 
 import (
 	"os"
+	"strconv"
 	"testing"
 )
 
-func TestDefaultMTU(t *testing.T) {
-	orig := os.Getenv("TS_DEBUG_MTU")
-	defer os.Setenv("TS_DEBUG_MTU", orig)
+func TestDefaultTunMTU(t *testing.T) {
+	// Save and restore the envknobs we will be changing
 
+	// The user can specify the MTU with this envknob.
+	ts_debug_mtu := os.Getenv("TS_DEBUG_MTU")
+	defer os.Setenv("TS_DEBUG_MTU", ts_debug_mtu)
 	os.Setenv("TS_DEBUG_MTU", "")
-	if DefaultMTU() != 1280 {
-		t.Errorf("DefaultMTU() = %d, want 1280", DefaultMTU())
+
+	// The intention is that PMTUD will be enabled by default someday.
+	ts_debug_pmtud := os.Getenv("TS_DEBUG_ENABLE_PMTUD")
+	defer os.Setenv("TS_DEBUG_ENABLE_PMTUD", ts_debug_pmtud)
+	os.Setenv("TS_DEBUG_ENABLE_PMTUD", "")
+
+	// With no MTU envknobs, we should get the conservative MTU.
+	if DefaultTunMTU() != safeTunMTU {
+		t.Errorf("DefaultTunMTU() = %d, want %d", DefaultTunMTU(), safeTunMTU)
+	}
+
+	// TS_DEBUG_MTU should take precedence over every other setting.
+	mtu := maxTunMTU - 1
+	os.Setenv("TS_DEBUG_MTU", strconv.Itoa(int(mtu)))
+	if DefaultTunMTU() != mtu {
+		t.Errorf("DefaultTunMTU() = %d, want %d, TS_DEBUG_MTU ignored", DefaultTunMTU(), mtu)
+	}
+
+	// MTU should be clamped to maxTunMTU.
+	mtu = maxTunMTU + 1
+	os.Setenv("TS_DEBUG_MTU", strconv.Itoa(int(mtu)))
+	if DefaultTunMTU() != maxTunMTU {
+		t.Errorf("DefaultTunMTU() = %d, want %d, clamping failed", DefaultTunMTU(), maxTunMTU)
 	}
 
-	os.Setenv("TS_DEBUG_MTU", "9000")
-	if DefaultMTU() != 9000 {
-		t.Errorf("DefaultMTU() = %d, want 9000", DefaultMTU())
+	// If PMTUD is enabled, the MTU should default to the largest
+	// probed MTU, but only if the user hasn't set an MTU.
+	os.Setenv("TS_DEBUG_MTU", "")
+	os.Setenv("TS_DEBUG_ENABLE_PMTUD", "true")
+	if DefaultTunMTU() != WireToTunMTU(MaxProbedWireMTU) {
+		t.Errorf("DefaultTunMTU() = %d, want %d", DefaultTunMTU(), WireToTunMTU(MaxProbedWireMTU))
+	}
+	// TS_DEBUG_MTU should take precedence over TS_DEBUG_ENABLE_PMTUD.
+	mtu = WireToTunMTU(MaxProbedWireMTU - 1)
+	os.Setenv("TS_DEBUG_MTU", strconv.Itoa(int(mtu)))
+	if DefaultTunMTU() != mtu {
+		t.Errorf("DefaultTunMTU() = %d, want %d", DefaultTunMTU(), mtu)
+	}
+}
+
+// Test the wire to Tailscale TUN MTU conversion corner cases
+func TestMTUConversion(t *testing.T) {
+	tests := []struct {
+		w WireMTU
+		t TunMTU
+	}{
+		{w: 0, t: 0},
+		{w: wgHeaderLen, t: 0},
+		{w: wgHeaderLen + 1, t: 1},
+		{w: 1360, t: 1280},
+		{w: 1500, t: 1420},
+		{w: 9000, t: 8920},
 	}
 
-	os.Setenv("TS_DEBUG_MTU", "123456789")
-	if DefaultMTU() != maxMTU {
-		t.Errorf("DefaultMTU() = %d, want %d", DefaultMTU(), maxMTU)
+	for _, tt := range tests {
+		m := WireToTunMTU(tt.w)
+		if m != tt.t {
+			t.Errorf("conversion of wire MTU %v to TUN MTU = %v, want %v", tt.w, m, tt.t)
+		}
+	}
+	for _, tt := range tests {
+		m := WireToTunMTU(tt.w)
```

Comment:
this block is identical to line 72's block? did you mean to test the other direction?

Reply:
Yes, thanks! Fixed.